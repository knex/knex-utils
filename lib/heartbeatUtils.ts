import Knex from 'knex'

export const HEARTBEAT_QUERIES = Object.freeze({
  ORACLE: 'select 1 from DUAL',
  POSTGRESQL: 'SELECT 1',
  MYSQL: 'SELECT 1',
  MSSQL: 'SELECT 1',
  SQLITE: 'SELECT 1',
  DEFAULT: 'SELECT 1',
})

export interface HeartbeatResult {
  isOk: boolean
  error?: Error
}

/**
 *
 * @param {Object} knex - Knex instance
 * @param {string} heartbeatQuery - SQL query that will be executed to check if connection is valid
 * @returns Promise<{Object|undefined}> wrapped error if connection is not valid, wrapped 'isOk: true' if it is valid, undefined if connection does not yet exist
 */
export function checkHeartbeat(
  knex: Knex,
  heartbeatQuery = HEARTBEAT_QUERIES.DEFAULT
): Promise<HeartbeatResult> {
  if (!knex) {
    throw new Error('Knex is mandatory parameter')
  }

  return knex
    .raw(heartbeatQuery)
    .then(() => {
      return {
        isOk: true,
      }
    })
    .catch((err: Error) => {
      return {
        isOk: false,
        error: err,
      }
    })
}

export type returnCallback<T> = (err?: any, result?: T) => any

export type CoinpaymentsRatesOpts = any
//export type returnCallback<T> = any
export type CoinpaymentsRatesResponse = any

export type CMDS = any

export class Something {
  public mapPayload<t>(x: any, y: any): any

  public rates(
    options?: CoinpaymentsRatesOpts | returnCallback<CoinpaymentsRatesResponse>,
    callback?: returnCallback
  ) {
    if (!options && !callback) {
      options = {}
    }
    if (typeof options === 'function') {
      callback = options
      options = {}
    }
    const requestPayload = mapPayload<CoinpaymentsRatesOpts>(options, {
      cmd: CMDS.RATES,
    })

    return {}
  }
}
