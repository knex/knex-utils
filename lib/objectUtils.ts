export function copyWithoutUndefined<T extends Record<K, V>, K extends string | number | symbol, V>(
  originalValue: T
): T {
  return Object.keys(originalValue).reduce((acc, key) => {
    // @ts-ignore
    if (originalValue[key] !== undefined) {
      // @ts-ignore
      acc[key] = originalValue[key]
    }
    return acc
  }, {} as Record<string, any>) as T
}

export function pick<T, K extends string | number | symbol>(
  source: T,
  names: readonly K[]
): Pick<T, Exclude<keyof T, Exclude<keyof T, K>>> {
  const result = {} as T
  let idx = 0
  while (idx < names.length) {
    if (names[idx] in source) {
      // @ts-ignore
      result[names[idx]] = source[names[idx]]
    }
    idx += 1
  }
  return result
}
