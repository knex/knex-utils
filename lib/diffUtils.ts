import { Knex } from 'knex'
import { pick } from './objectUtils'
import { chunk } from './chunkUtils'

function arrayIncludesWith(array: any[], target: any, comparator: any) {
  if (array == null) {
    return false
  }

  for (const value of array) {
    if (comparator(target, value)) {
      return true
    }
  }
  return false
}

function baseDifference<T>(initialArray: T[], newArray: T[], comparator: any): EntityListDiff<T> {
  const includes = arrayIncludesWith
  const newEntries: T[] = []
  const removedEntries: T[] = []

  for (const value of initialArray) {
    if (!includes(newArray, value, comparator)) {
      removedEntries.push(value)
    }
  }

  for (const value of newArray) {
    if (!includes(initialArray, value, comparator)) {
      newEntries.push(value)
    }
  }

  return {
    removedEntries,
    newEntries,
  }
}

export type EntityListDiff<T> = {
  newEntries: T[]
  removedEntries: T[]
}

export function calculateEntityListDiff<T>(
  oldList: T[],
  newList: T[],
  idFields: string[]
): EntityListDiff<T> {
  const comparator = (value1: any, value2: any) => {
    for (const idField of idFields) {
      if (value1[idField] !== value2[idField]) {
        return false
      }
    }
    return true
  }

  return baseDifference(oldList, newList, comparator)
}

export type UpdateJoinTableParams = {
  filterCriteria: Record<string, any>
  table: string
  primaryKeyField?: string
  chunkSize?: number
  idFields: string[]
  transactionProvider?: Knex.TransactionProvider
  transaction?: Knex.Transaction
}

async function getKnexOrTrx(
  knex: Knex,
  params: UpdateJoinTableParams
): Promise<Knex | Knex.Transaction> {
  if (knex.client.driverName === 'sqlite3') {
    return knex
  }

  if (params.transaction) {
    return params.transaction
  }

  if (params.transactionProvider) {
    return params.transactionProvider()
  }

  return knex.transaction()
}

export async function updateJoinTable<T>(
  knex: Knex,
  newList: T[],
  params: UpdateJoinTableParams
): Promise<EntityListDiff<T>> {
  const chunkSize = params.chunkSize || 100
  const trx = await getKnexOrTrx(knex, params)
  try {
    const oldList = await trx(params.table).select('*').where(params.filterCriteria)
    const diff = calculateEntityListDiff(oldList, newList, params.idFields)

    const insertChunks = chunk(diff.newEntries, chunkSize)
    for (const insertChunk of insertChunks) {
      await trx(params.table).insert(insertChunk)
    }

    // If we have a primary key, then we can delete in batch
    if (params.primaryKeyField) {
      const deleteIds = diff.removedEntries.map((entry) => {
        return entry[params.primaryKeyField!]
      })
      const deleteChunks = chunk(deleteIds, chunkSize)
      for (const deleteChunk of deleteChunks) {
        await trx(params.table).delete().whereIn(params.primaryKeyField, deleteChunk)
      }
    }
    // Otherwise we have to delete one-by-one
    else {
      const deleteCriteria = diff.removedEntries.map((entry) => {
        return pick(entry, params.idFields)
      })
      for (const entry of deleteCriteria) {
        await trx(params.table).delete().where(entry)
      }
    }

    if (trx.isTransaction && !params.transaction && !params.transactionProvider) {
      await (trx as Knex.Transaction).commit()
    }
    return diff
  } catch (err) {
    if (trx.isTransaction) {
      await (trx as Knex.Transaction).rollback()
    }
    throw err
  }
}
