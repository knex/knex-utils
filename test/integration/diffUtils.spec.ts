import { Knex } from 'knex'
import { getAllDbs, getKnexForDb } from './helpers/knexInstanceProvider'
import { createJoinTable, dropJoinTable } from './helpers/tableCreator'
import { updateJoinTable } from '../../lib/diffUtils'
import { generateAssets } from './helpers/entityGenerator'

describe('diffUtils integration', () => {
  getAllDbs().forEach((db) => {
    describe(db, () => {
      let knex: Knex
      beforeEach(async () => {
        knex = getKnexForDb(db)
        await createJoinTable(knex)
      })

      afterEach(async () => {
        await dropJoinTable(knex)
        await knex.destroy()
      })

      describe('updateJoinTable', () => {
        it('correctly inserts new entries', async () => {
          const newList = generateAssets(0, { orgId: 'kiberion', linkType: 'primaryAsset' }, 10)
          await updateJoinTable(knex, newList, {
            primaryKeyField: 'id',
            idFields: ['userId', 'orgId', 'linkType'],
            table: 'joinTable',
            chunkSize: 3,
            filterCriteria: {
              orgId: 'kiberion',
              linkType: 'primaryAsset',
            },
          })

          const result = await knex('joinTable').select('*')
          expect(result).toMatchSnapshot()
        })
      })
    })
  })
})
