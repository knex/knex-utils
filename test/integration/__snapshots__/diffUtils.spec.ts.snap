// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diffUtils integration mssql updateJoinTable correctly inserts new entries 1`] = `
Array [
  Object {
    "id": 1,
    "linkType": "primaryAsset",
    "name": "2",
    "orgId": "kiberion",
    "userId": "1",
  },
  Object {
    "id": 2,
    "linkType": "primaryAsset",
    "name": "4",
    "orgId": "kiberion",
    "userId": "3",
  },
  Object {
    "id": 3,
    "linkType": "primaryAsset",
    "name": "6",
    "orgId": "kiberion",
    "userId": "5",
  },
  Object {
    "id": 4,
    "linkType": "primaryAsset",
    "name": "8",
    "orgId": "kiberion",
    "userId": "7",
  },
  Object {
    "id": 5,
    "linkType": "primaryAsset",
    "name": "10",
    "orgId": "kiberion",
    "userId": "9",
  },
  Object {
    "id": 6,
    "linkType": "primaryAsset",
    "name": "12",
    "orgId": "kiberion",
    "userId": "11",
  },
  Object {
    "id": 7,
    "linkType": "primaryAsset",
    "name": "14",
    "orgId": "kiberion",
    "userId": "13",
  },
  Object {
    "id": 8,
    "linkType": "primaryAsset",
    "name": "16",
    "orgId": "kiberion",
    "userId": "15",
  },
  Object {
    "id": 9,
    "linkType": "primaryAsset",
    "name": "18",
    "orgId": "kiberion",
    "userId": "17",
  },
  Object {
    "id": 10,
    "linkType": "primaryAsset",
    "name": "20",
    "orgId": "kiberion",
    "userId": "19",
  },
]
`;

exports[`diffUtils integration mysql updateJoinTable correctly inserts new entries 1`] = `
Array [
  RowDataPacket {
    "id": 1,
    "linkType": "primaryAsset",
    "name": "2",
    "orgId": "kiberion",
    "userId": "1",
  },
  RowDataPacket {
    "id": 2,
    "linkType": "primaryAsset",
    "name": "4",
    "orgId": "kiberion",
    "userId": "3",
  },
  RowDataPacket {
    "id": 3,
    "linkType": "primaryAsset",
    "name": "6",
    "orgId": "kiberion",
    "userId": "5",
  },
  RowDataPacket {
    "id": 4,
    "linkType": "primaryAsset",
    "name": "8",
    "orgId": "kiberion",
    "userId": "7",
  },
  RowDataPacket {
    "id": 5,
    "linkType": "primaryAsset",
    "name": "10",
    "orgId": "kiberion",
    "userId": "9",
  },
  RowDataPacket {
    "id": 6,
    "linkType": "primaryAsset",
    "name": "12",
    "orgId": "kiberion",
    "userId": "11",
  },
  RowDataPacket {
    "id": 7,
    "linkType": "primaryAsset",
    "name": "14",
    "orgId": "kiberion",
    "userId": "13",
  },
  RowDataPacket {
    "id": 8,
    "linkType": "primaryAsset",
    "name": "16",
    "orgId": "kiberion",
    "userId": "15",
  },
  RowDataPacket {
    "id": 9,
    "linkType": "primaryAsset",
    "name": "18",
    "orgId": "kiberion",
    "userId": "17",
  },
  RowDataPacket {
    "id": 10,
    "linkType": "primaryAsset",
    "name": "20",
    "orgId": "kiberion",
    "userId": "19",
  },
]
`;

exports[`diffUtils integration mysql2 updateJoinTable correctly inserts new entries 1`] = `
Array [
  TextRow {
    "id": 1,
    "linkType": "primaryAsset",
    "name": "2",
    "orgId": "kiberion",
    "userId": "1",
  },
  TextRow {
    "id": 2,
    "linkType": "primaryAsset",
    "name": "4",
    "orgId": "kiberion",
    "userId": "3",
  },
  TextRow {
    "id": 3,
    "linkType": "primaryAsset",
    "name": "6",
    "orgId": "kiberion",
    "userId": "5",
  },
  TextRow {
    "id": 4,
    "linkType": "primaryAsset",
    "name": "8",
    "orgId": "kiberion",
    "userId": "7",
  },
  TextRow {
    "id": 5,
    "linkType": "primaryAsset",
    "name": "10",
    "orgId": "kiberion",
    "userId": "9",
  },
  TextRow {
    "id": 6,
    "linkType": "primaryAsset",
    "name": "12",
    "orgId": "kiberion",
    "userId": "11",
  },
  TextRow {
    "id": 7,
    "linkType": "primaryAsset",
    "name": "14",
    "orgId": "kiberion",
    "userId": "13",
  },
  TextRow {
    "id": 8,
    "linkType": "primaryAsset",
    "name": "16",
    "orgId": "kiberion",
    "userId": "15",
  },
  TextRow {
    "id": 9,
    "linkType": "primaryAsset",
    "name": "18",
    "orgId": "kiberion",
    "userId": "17",
  },
  TextRow {
    "id": 10,
    "linkType": "primaryAsset",
    "name": "20",
    "orgId": "kiberion",
    "userId": "19",
  },
]
`;

exports[`diffUtils integration postgres updateJoinTable correctly inserts new entries 1`] = `
Array [
  Object {
    "id": 1,
    "linkType": "primaryAsset",
    "name": "2",
    "orgId": "kiberion",
    "userId": "1",
  },
  Object {
    "id": 2,
    "linkType": "primaryAsset",
    "name": "4",
    "orgId": "kiberion",
    "userId": "3",
  },
  Object {
    "id": 3,
    "linkType": "primaryAsset",
    "name": "6",
    "orgId": "kiberion",
    "userId": "5",
  },
  Object {
    "id": 4,
    "linkType": "primaryAsset",
    "name": "8",
    "orgId": "kiberion",
    "userId": "7",
  },
  Object {
    "id": 5,
    "linkType": "primaryAsset",
    "name": "10",
    "orgId": "kiberion",
    "userId": "9",
  },
  Object {
    "id": 6,
    "linkType": "primaryAsset",
    "name": "12",
    "orgId": "kiberion",
    "userId": "11",
  },
  Object {
    "id": 7,
    "linkType": "primaryAsset",
    "name": "14",
    "orgId": "kiberion",
    "userId": "13",
  },
  Object {
    "id": 8,
    "linkType": "primaryAsset",
    "name": "16",
    "orgId": "kiberion",
    "userId": "15",
  },
  Object {
    "id": 9,
    "linkType": "primaryAsset",
    "name": "18",
    "orgId": "kiberion",
    "userId": "17",
  },
  Object {
    "id": 10,
    "linkType": "primaryAsset",
    "name": "20",
    "orgId": "kiberion",
    "userId": "19",
  },
]
`;

exports[`diffUtils integration sqlite3 updateJoinTable correctly inserts new entries 1`] = `
Array [
  Object {
    "id": 1,
    "linkType": "primaryAsset",
    "name": "2",
    "orgId": "kiberion",
    "userId": "1",
  },
  Object {
    "id": 2,
    "linkType": "primaryAsset",
    "name": "4",
    "orgId": "kiberion",
    "userId": "3",
  },
  Object {
    "id": 3,
    "linkType": "primaryAsset",
    "name": "6",
    "orgId": "kiberion",
    "userId": "5",
  },
  Object {
    "id": 4,
    "linkType": "primaryAsset",
    "name": "8",
    "orgId": "kiberion",
    "userId": "7",
  },
  Object {
    "id": 5,
    "linkType": "primaryAsset",
    "name": "10",
    "orgId": "kiberion",
    "userId": "9",
  },
  Object {
    "id": 6,
    "linkType": "primaryAsset",
    "name": "12",
    "orgId": "kiberion",
    "userId": "11",
  },
  Object {
    "id": 7,
    "linkType": "primaryAsset",
    "name": "14",
    "orgId": "kiberion",
    "userId": "13",
  },
  Object {
    "id": 8,
    "linkType": "primaryAsset",
    "name": "16",
    "orgId": "kiberion",
    "userId": "15",
  },
  Object {
    "id": 9,
    "linkType": "primaryAsset",
    "name": "18",
    "orgId": "kiberion",
    "userId": "17",
  },
  Object {
    "id": 10,
    "linkType": "primaryAsset",
    "name": "20",
    "orgId": "kiberion",
    "userId": "19",
  },
]
`;
